### ISSUE TITLE
A lot of type errors

### ISSUE THREAD

**Issue by InCogNiTo124** [georgealways/lil-gui#93]
tsc version 4.9.5
lil-gui version 0.18.0

There now seems to be a lot more errors than it used to be regarding types:

<details>
<summary>Errors</summary>

```
node_modules/lil-gui/dist/lil-gui.esm.d.ts:305:5 - error TS2416: Property 'updateDisplay' in type 'BooleanController' is not assignable to the same property in base type 'Controller'.
  Type '() => BooleanController' is not assignable to type '() => this'.
    Type 'BooleanController' is not assignable to type 'this'.
      'BooleanController' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'BooleanController'.

305     updateDisplay(): BooleanController;
        ~~~~~~~~~~~~~

node_modules/lil-gui/dist/lil-gui.esm.d.ts:336:5 - error TS2416: Property 'reset' in type 'ColorController' is not assignable to the same property in base type 'Controller'.
  Type '() => ColorController' is not assignable to type '() => this'.
    Type 'ColorController' is not assignable to type 'this'.
      'ColorController' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'ColorController'.

336     reset(): ColorController;
        ~~~~~

node_modules/lil-gui/dist/lil-gui.esm.d.ts:340:5 - error TS2416: Property 'updateDisplay' in type 'ColorController' is not assignable to the same property in base type 'Controller'.
  Type '() => ColorController' is not assignable to type '() => this'.
    Type 'ColorController' is not assignable to type 'this'.
      'ColorController' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'ColorController'.

340     updateDisplay(): ColorController;
        ~~~~~~~~~~~~~

node_modules/lil-gui/dist/lil-gui.esm.d.ts:613:5 - error TS2416: Property 'decimals' in type 'NumberController' is not assignable to the same property in base type 'Controller'.
  Type '(decimals: any) => NumberController' is not assignable to type '(decimals: number) => this'.
    Type 'NumberController' is not assignable to type 'this'.
      'NumberController' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'NumberController'.

613     decimals(decimals: any): NumberController;
        ~~~~~~~~

node_modules/lil-gui/dist/lil-gui.esm.d.ts:615:5 - error TS2416: Property 'min' in type 'NumberController' is not assignable to the same property in base type 'Controller'.
  Type '(min: any) => NumberController' is not assignable to type '(min: number) => this'.
    Type 'NumberController' is not assignable to type 'this'.
      'NumberController' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'NumberController'.

615     min(min: any): NumberController;
        ~~~

node_modules/lil-gui/dist/lil-gui.esm.d.ts:617:5 - error TS2416: Property 'max' in type 'NumberController' is not assignable to the same property in base type 'Controller'.
  Type '(max: any) => NumberController' is not assignable to type '(max: number) => this'.
    Type 'NumberController' is not assignable to type 'this'.
      'NumberController' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'NumberController'.

617     max(max: any): NumberController;
        ~~~

node_modules/lil-gui/dist/lil-gui.esm.d.ts:619:5 - error TS2416: Property 'step' in type 'NumberController' is not assignable to the same property in base type 'Controller'.
  Type '(step: any, explicit?: boolean | undefined) => NumberController' is not assignable to type '(step: number) => this'.
    Type 'NumberController' is not assignable to type 'this'.
      'NumberController' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'NumberController'.

619     step(step: any, explicit?: boolean): NumberController;
        ~~~~

node_modules/lil-gui/dist/lil-gui.esm.d.ts:622:5 - error TS2416: Property 'updateDisplay' in type 'NumberController' is not assignable to the same property in base type 'Controller'.
  Type '() => NumberController' is not assignable to type '() => this'.
    Type 'NumberController' is not assignable to type 'this'.
      'NumberController' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'NumberController'.

622     updateDisplay(): NumberController;
        ~~~~~~~~~~~~~

node_modules/lil-gui/dist/lil-gui.esm.d.ts:649:5 - error TS2416: Property 'updateDisplay' in type 'OptionController' is not assignable to the same property in base type 'Controller'.
  Type '() => OptionController' is not assignable to type '() => this'.
    Type 'OptionController' is not assignable to type 'this'.
      'OptionController' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'OptionController'.

649     updateDisplay(): OptionController;
        ~~~~~~~~~~~~~

node_modules/lil-gui/dist/lil-gui.esm.d.ts:655:5 - error TS2416: Property 'updateDisplay' in type 'StringController' is not assignable to the same property in base type 'Controller'.
  Type '() => StringController' is not assignable to type '() => this'.
    Type 'StringController' is not assignable to type 'this'.
      'StringController' is assignable to the constraint of type 'this', but 'this' could be instantiated with a different subtype of constraint 'StringController'.

655     updateDisplay(): StringController;
        ~~~~~~~~~~~~~


Found 10 errors in the same file, starting at: node_modules/lil-gui/dist/lil-gui.esm.d.ts:305

```

</details>

Since my code hasn't changed (this was a dep bump from 0.17.0 to 0.18.0, and typescript from 4.9.4 to 4.9.5) I believe this has something to do with types

**Comment by georgealways** [comment]
Definitely seems like it, thanks for calling this out. I'm not a typescript user so you'll have to help me a bit. 

So nothing changed about the type definition script https://github.com/georgealways/lil-gui/blob/master/package.json#L27

But there's definitely more changes between the .d.ts files than I would expect between 0.17 and 0.18: https://www.diffchecker.com/C6r3SEEK/

The only real sanity check I do on the d.ts file before I publish is to import lil-gui in a test.ts file and see if VSCode can describe the parameters in `new GUI`. Is there a command I can run to get the sort of output you're seeing above?

**Comment by InCogNiTo124** [comment]
I'm just learning typescript myself so I don't think I have the experience I need :sweat_smile: 

as for the repro, I experienced this problem when building a private project https://github.com/InCogNiTo124/kepler-orbit. You can clone it and try it yourself

This is the command I've been running:
```
rm -rf node_modules/ yarn.lock && yarn install && node_modules/typescript/bin/tsc
```

- **master**: command works
- **renovate/all**: command fails


---

I now see vite is also updated, from 4.0.3 to 4.1.4, but this should not make difference, I guess

**Comment by Methuselah96** [comment]
> Is there a command I can run to get the sort of output you're seeing above?

With https://github.com/georgealways/lil-gui/pull/94 you can run `npm run type-check` after the types have been generated and it will produce these errors.

**Comment by Methuselah96** [comment]
@georgealways Yes, I also created https://github.com/georgealways/lil-gui/pull/95, maybe you missed that.

**Comment by georgealways** [comment]
missed it entirely, thanks :)

**Comment by InCogNiTo124** [comment]
Just to clarify, you really think this may be a bug due to typescript itself? what are the potential steps if it's not?

**Comment by Methuselah96** [comment]
I think the plan is to lock TypeScript to 4.5 (https://github.com/georgealways/lil-gui/pull/96) since that is @georgealways's [preferred solution](https://github.com/georgealways/lil-gui/pull/95#issuecomment-1450628105) out of the possible solutions we've come up with so far.

**Comment by georgealways** [comment]
Actually keeping the "help wanted" here—maybe someone knows a way to get the output we want from the latest version of typescript. Ideally #96 is a temporary solve.

I'll be publishing version 0.18.1 with new type definitions created using typescript@4.5, which we know works.

**Comment by georgealways** [comment]
@Methuselah96 @InCogNiTo124 Just published 0.18.1 with the typescript fixes. Let me know if that's working. Thanks for your help!

**Comment by InCogNiTo124** [comment]
@georgealways The fix has solved the errors with `lil-gui` thank you!
Now I have some other bugs with three.js :sweat_smile:
